optimizely/odp/lru_cache.py

```diff
@@ -0,0 +1,133 @@
+# Copyright 2022, Optimizely
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+
+from __future__ import annotations
+from dataclasses import dataclass, field
+import threading
+from time import time
+from collections import OrderedDict
+from typing import Optional, Generic, TypeVar, Hashable
+from sys import version_info
+
+if version_info < (3, 8):
+    from typing_extensions import Protocol
+else:
+    from typing import Protocol  # type: ignore
+
+
+# generic type definitions for LRUCache parameters
+K = TypeVar('K', bound=Hashable, contravariant=True)
+V = TypeVar('V')
+
+
+class LRUCache(Generic[K, V]):
+    """Least Recently Used cache that invalidates entries older than the timeout."""
+
+    def __init__(self, capacity: int, timeout_in_secs: float):
+        self.lock = threading.Lock()
+        self.map: OrderedDict[K, CacheElement[V]] = OrderedDict()
+        self.capacity = capacity
+        self.timeout = timeout_in_secs
+
+    def lookup(self, key: K) -> Optional[V]:
+        """Return the non-stale value associated with the provided key and move the
+        element to the end of the cache. If the selected value is stale, remove it from
+        the cache and clear the entire cache if stale.
+        """
+        if self.capacity <= 0:
+            return None
+
+        with self.lock:
+            if key not in self.map:
+                return None
+
+            self.map.move_to_end(key)
+            element = self.map[key]
+
+            if element._is_stale(self.timeout):
+                del self.map[key]
+                if self._all_stale():
+                    self.map.clear()
+                return None
+
+        return element.value
+
+    def save(self, key: K, value: V) -> None:
```

Comment:
Just a note that Jae/Swift have this function async. Andy and I discussed the concurrency in Py will be handled at odp segment manager level --> at least for the graphQL fetching. But I think also for this. I think we discussed this. Any thoughts?